src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2283 @@
+                                RELAY_COMMAND_DROP, NULL, 0);
+  } else {
+    // If we're a non-origin circ, we can just send from here as if we're the
+    // edge.
+    relay_send_command_from_edge(0, mi->on_circ, RELAY_COMMAND_DROP, NULL,
+                                 0, NULL);
+  }
+
+  rep_hist_padding_count_write(PADDING_TYPE_DROP);
+  /* This is a padding cell sent from the client or from the middle node,
+   * (because it's invoked from circuitpadding.c) */
+  circpad_cell_event_padding_sent(circ);
+
+  /* The circpad_cell_event_padding_sent() could cause us to transition.
+   * Check that we still have a padding machineinfo, and then check our token
+   * supply. */
+  if (circ->padding_info[machine_idx] != NULL)
+    circpad_check_token_supply(circ->padding_info[machine_idx]);
+}
+
+/**
+ * Tor-timer compatible callback that tells us to send a padding cell.
+ *
+ * Conver the handle to a pointer; do some basic sanity checks, then
+ * send appropriate data to the callback helper to send the cell.
+ */
+static void
+circpad_send_padding_callback(tor_timer_t *timer, void *args,
+                              const struct monotime_t *time)
+{
+  circpad_machineinfo_t *mi =
+    circpad_machineinfo_handle_get((struct circpad_machineinfo_handle_t*)args);
+  (void)timer; (void)time;
+
+  if (mi && mi->on_circ) {
+    assert_circuit_ok(mi->on_circ);
+    circpad_send_padding_cell_for_callback(mi);
+  } else {
+    // This shouldn't happen (represents a handle leak)
+    log_fn(LOG_WARN,LD_CIRC,
+            "Circuit closed while waiting for padding timer.");
+    tor_fragile_assert();
+  }
+
+  // TODO-MP-AP: Unify this counter with channelpadding for rephist stats
+  //total_timers_pending--;
+}
+
+/**
+ * Cache our consensus parameters upon consensus update.
+ */
+void
+circpad_new_consensus_params(networkstatus_t *ns)
+{
+  circpad_per_hop_allowed_cells =
+      networkstatus_get_param(ns, "circpad_allowed_cells",
+         0, 0, UINT16_MAX-1);
+
+  circpad_per_hop_max_padding_percent =
+      networkstatus_get_param(ns, "circpad_max_padding_pct",
+         0, 0, 100);
+}
+
+/**
+ * Check this machine against its padding limits, as well as global
+ * consensus limits.
+ *
+ * We have two limits: a percent and a cell count. The cell count
+ * limit must be reached before the percent is enforced (this is to
+ * optionally allow very light padding of things like circuit setup
+ * while there is no other traffic on the circuit).
+ *
+ * The consensus versions are per-hop values, so they must be divided
+ * by target_hopnum.
+ *
+ * Returns 1 if limits are set and we've hit them. Otherwise returns 0.
+ */
+static bool
+circpad_machine_reached_padding_limit(circpad_machineinfo_t *mi)
+{
+  const circpad_machine_t *machine = CIRCPAD_GET_MACHINE(mi);
+  uint8_t max_padding_pct;
+  uint16_t allowed_cells;
+
+  /* If consensus parameters have been set, they override the per-machine
```

Comment:
Similar as for being able to specify MiddleNodes, it would be useful to be able to override consensus limits for sake of research.

Reply:
Hrmm I simplified this. I don't think we actually need consensus machine limits, since presumably any machines we accept into the consensus should have their own machine limit.

I changed the consensus limit to a global limit in db6e20fd573f0a4111fdba273fece6cd347bd240 and put a TODO in the source for them not to apply to machines loaded from torrc (which we have not implemented yet).